### PR TITLE
Multigrid with 3D implicit line agglomeration

### PR DESCRIPTION
**Pull request by bigfooted**
## Proposed Changes

*fix implicit line agglomeration*

When MG_IMPLICIT_LINE_AGGLOMERATION=YES, we try to create structured coarser meshes by first agglomerating the boundary, then use a paving method, also known as an advancing front method. We advance the entire front layer by layer, so all viscous walls advance one layer at a time, until the end, until an edge direction deviates more than 30 degrees or until a collision. When paving for a single cell fails, paving for the entire marker stops.

After agglomeration of the viscous walls, the other boundaries are agglomerated in a structured way if their first cell aspect ratio is higher than 2. For structured quadrilateral meshes with inflation layers this method leads to very high quality multigrids.
With MPI, the layers try to continue through the mpi interface. When the mpi interface created by parmetis is not very 'straight', this method often terminates and quality deteriorates. However, even the fallback to regular agglomeration still leads to high quality meshes.

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [x] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [x] I used the pre-commit hook to prevent dirty commits and used `pre-commit run --all` to format old commits.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

### CODE REVIEW

File: SU2_CFD/src/integration/CMultiGridIntegration.cpp [hunk 1]
```diff
@@ -814,7 +827,7 @@ void CMultiGridIntegration::SetProlongated_Correction(CSolver *sol_fine, CGeomet
     for (auto iVar = 0u; iVar < nVar; iVar++) {
       /*--- Prevent a fine grid divergence due to a coarse grid divergence ---*/
       if (Residual_Fine[iVar] != Residual_Fine[iVar])
-        Residual_Fine[iVar] = 0.0;
+Residual_Fine[iVar] = 0.0;
```
Comment (bigfooted) [inline SU2_CFD/src/integration/CMultiGridIntegration.cpp]:
```suggestion
        Residual_Fine[iVar] = 0.0;
```

**Comment by bigfooted** [comment]
screenshot of the first level agglomeration for the turbulent flat plate. the black lines is the original mesh, the red lines is the new agglomerated mesh. Note that boundary nodes are not agglomerated with interior nodes. 
image

**Comment by bigfooted** [comment]
3D bump in channel, vertical slice through the domain:
image

**Comment by bigfooted** [comment]
This image is a close-up of the slice through the 3D bump in channel. The orange line shows the location of the MPI interface. Left and right of the bump the wall becomes a symmetry plane, and at this interface there is no horizontal agglomeration. 
image

**Comment by bigfooted** [comment]
note that most of the agglomerations where no horizontal agglomeration occurs is due to some MPI collision. without MPI, the agglomeration looks better:
image

This shows that improving parmetis MPI agglomeration can result in better multigrid agglomeration

File: Common/include/geometry/CMultiGridGeometry.hpp [hunk 2]
```diff
@@ -78,16 +87,71 @@ class CMultiGridGeometry final : public CGeometry {
   su2double ComputeLocalCurvature(const CGeometry* fine_grid, unsigned long iPoint, unsigned short iMarker) const;
 
   /*!
-   * \brief Agglomerate high-aspect-ratio interior cells along implicit lines from wall vertices.
+   * \brief Pave the domain with advancing fronts rising from the boundary patches.
```
Comment (bigfooted) [inline Common/include/geometry/CMultiGridGeometry.hpp]:
```suggestion
   * \brief Pave the domain with advancing fronts extruded from the boundary patches.
```

File: Common/src/geometry/CMultiGridGeometry.cpp [hunk 3]
```diff
@@ -88,11 +88,40 @@ CMultiGridGeometry::CMultiGridGeometry(CGeometry* fine_grid, CConfig* config, un
...
+  vector<char> onPhysBoundary(fine_grid->GetnPoint(), 0);
+  for (auto iMarker = 0u; iMarker < fine_grid->GetnMarker(); iMarker++) {
+    if (config->GetMarker_All_KindBC(iMarker) == SEND_RECEIVE) continue;
+    for (auto iVertex = 0ul; iVertex < fine_grid->GetnVertex(iMarker); iVertex++)
+      onPhysBoundary[fine_grid->vertex[iMarker][iVertex]->GetNode()] = 1;
+  }
+  /*--- Nodes where two different boundary conditions meet. The rule has to hold for every phase or
+   *    the same node is treated one way by the paving and another here. In 2D the corner test below
+   *    already refused them; in 3D a ridge of such nodes carries one identical marker PAIR all along
+   *    it and would otherwise pair up with itself quite happily. ---*/
```
Comment (bigfooted) [inline Common/src/geometry/CMultiGridGeometry.cpp]:
```suggestion
  /*--- Nodes where two different boundary conditions meet.  ---*/
```

File: Common/src/geometry/CMultiGridGeometry.cpp [hunk 4]
```diff
@@ -88,11 +88,40 @@ CMultiGridGeometry::CMultiGridGeometry(CGeometry* fine_grid, CConfig* config, un
...
+   *    CVs it creates occupy [firstLineCV, endLineCV), which is how the repair passes identify them. ---*/
+  const auto firstLineCV = Index_CoarseCV;
+  if (config->GetMGOptions().MG_Implicit_Lines) {
+    pavingReport = AgglomerateImplicitLines(Index_CoarseCV, fine_grid, config, iMesh);
+  }
+  const auto endLineCV = Index_CoarseCV;
+
+  /*--- Points carrying a physical boundary condition. SEND_RECEIVE is not one: it only records that
+   *    the point is mirrored on another rank. Used by the repair passes below to tell which coarse
+   *    CVs touch a boundary, see cvOnBoundary. ---*/
```
Comment (bigfooted) [inline Common/src/geometry/CMultiGridGeometry.cpp]:
```suggestion
  /*--- Points carrying a physical boundary condition. This does not include SEND_RECEIVE. ---*/
```

File: Common/src/geometry/CMultiGridGeometry.cpp [hunk 5]
```diff
@@ -118,13 +147,18 @@ CMultiGridGeometry::CMultiGridGeometry(CGeometry* fine_grid, CConfig* config, un
         unsigned short copy_marker[3] = {};
         marker_seed.push_back(iMarker);
 
-        /*--- For a particular point in the fine grid we save all the markers
-         that are in that point ---*/
+        /*--- For a particular point in the fine grid we save all the physical markers that are in
+         that point. SEND_RECEIVE markers are deliberately not counted: including them would make
+         an ordinary wall point look like a ridge, and a wall/symmetry ridge look like a corner
+         (which the counter > 2 rule below then refuses to agglomerate at all), so a point would be
+         classified differently depending only on where the partition happens to cut. ---*/
```
Comment (bigfooted) [inline Common/src/geometry/CMultiGridGeometry.cpp]:
```suggestion
        /*--- For a particular point in the fine grid we save all the physical markers that are in
         that point.  ---*/
```

File: Common/src/geometry/CMultiGridGeometry.cpp [hunk 6]
```diff
@@ -163,23 +197,21 @@ CMultiGridGeometry::CMultiGridGeometry(CGeometry* fine_grid, CConfig* config, un
...
+          /*--- In 2D that is a genuine corner in the geometry, which is never agglomerated. A wall
+                point merely split by a partition interface no longer reaches this branch: it counts
+                one physical marker and is handled above as the valley point it is. ---*/
+          if (nDim == 2) agglomerate_seed = false;
+          /*--- In 3D, this is a ridge point (an edge feature where two surface markers meet).
+                Always allow it to attempt agglomeration here; SetBoundAgglomeration() enforces
+                the actual ridge-ridge rule downstream: it may only pair with a neighboring ridge
+                point that carries the identical physical marker pair. A mismatched marker pair
+                usually indicates a genuine sharp corner in the geometry and is correctly left
+                un-merged (falls through to the singleton leftover loop). ---*/
```
Comment (bigfooted) [inline Common/src/geometry/CMultiGridGeometry.cpp]:
```suggestion
          /*--- In 3D, this is a ridge point (an edge where two surface markers meet). ---*/
```

File: Common/src/geometry/CMultiGridGeometry.cpp [hunk 7]
```diff
@@ -231,12 +268,15 @@ CMultiGridGeometry::CMultiGridGeometry(CGeometry* fine_grid, CConfig* config, un
...
             }
           }
 
-          /*--- Only take into account indirect neighbors for 3D faces, not 2D. ---*/
-          if (nDim == 3) {
+          /*--- Indirect neighbors only for 3D faces. The size test is needed on entry too: the sweep
+           *    above leaves the CV exactly full when it hit the limit, and without this the first
+           *    indirect candidate pushed it past the limit, after which the equality test never matched
+           *    and the CV grew unbounded. ---*/
+          if ((nDim == 3) && (nChildren < maxAgglomSize)) {
```
Comment (bigfooted) [inline Common/src/geometry/CMultiGridGeometry.cpp]:
```suggestion
          /*--- Indirect neighbors only for 3D faces.  ---*/
          if ((nDim == 3) && (nChildren < maxAgglomSize)) {
```

File: Common/src/geometry/CMultiGridGeometry.cpp [hunk 8]
```diff
@@ -311,12 +356,23 @@ CMultiGridGeometry::CMultiGridGeometry(CGeometry* fine_grid, CConfig* config, un
...
-  /*--- Agglomerate high-aspect-ratio interior nodes along implicit lines from walls. ---*/
-  if (config->GetMGOptions().MG_Implicit_Lines) {
-    AgglomerateImplicitLines(Index_CoarseCV, fine_grid, config, MGQueue_InnerCV);
-  }
+  /*--- STEP 2: Agglomerate the domain points. A seed grows one node at a time, taking the candidate
+   *    that shares the most edges with the current members so the CV closes into a block rather than a
+   *    star. Ties break on distance to the centroid measured in cells, using a local frame built from
+   *    the seed's own edges, so a stretched cell does not pull the CV along the boundary layer. ---*/
+
+  /*--- Scratch shared by all seeds. The markers are cleared per CV, touching only what was used. ---*/
```
Comment (bigfooted) [inline Common/src/geometry/CMultiGridGeometry.cpp]:
```suggestion
  /*--- STEP 2: Agglomerate the domain points. A seed grows one node at a time, taking the candidate
   *    that shares the most edges with the current members. ---*/
```

File: Common/src/geometry/CMultiGridGeometry.cpp [hunk 9]
```diff
@@ -311,12 +356,23 @@ CMultiGridGeometry::CMultiGridGeometry(CGeometry* fine_grid, CConfig* config, un
...
+  vector<char> inCV(fine_grid->GetnPoint(), 0);
+  vector<char> isCandidate(fine_grid->GetnPoint(), 0);
+  vector<unsigned long> members, candidates;
+  members.reserve(maxAgglomSize);
 
-  /*--- STEP 2: Agglomerate the domain points. ---*/
+  /*--- A local frame at the seed: up to nDim of its incident edges, as mutually orthogonal as possible,
+   *    each with its own length. An offset is resolved onto these and each component divided by that
+   *    direction's spacing, so the measure turns with the mesh and stays correct on a curved boundary. ---*/
```
Comment (bigfooted) [inline Common/src/geometry/CMultiGridGeometry.cpp]:
```suggestion
  /*--- A local frame at the seed: up to nDim of its incident edges, as mutually orthogonal as possible,
   *    each with its own length. An offset is resolved onto these and each component divided by that
   *    direction's spacing. ---*/
```

**Comment by bigfooted** [comment]
Code is still a bit bloated, I will do a second pass...
I will also modify some testcase setups so they have settings that converge well. Some testcases you can increase CFL from 10 -> 1000 without any problem.

**Comment by bigfooted** [comment]
RAE2822 airfoil:
image
and a zoom:
image

**Comment by bigfooted** [comment]
RAE2822 case
convergence with implicit lines: 185 iterations, 19 s.
image

single grid: 740 iterations, 29 s. CFL had to be reduced from 1000 -> 100 to converge it. And it still has some weird convergence peaks at the end
image

File: q
Comment (pcarruscag) [inline q:1]: Something weird happened here?

**Review by pcarruscag — APPROVED** [review]

File: q
Comment (bigfooted) [reply q:1]: yes, I removed it, probably me aggresively using vi :wq but then writing a file named q. :-)
